rfcs/015 - TCP RFC.md

```diff
@@ -0,0 +1,196 @@
+
+While changes to the language are on the table (and some of the alternatives will cover a few of them), the best overall solution (considering implementation complexity and cognitive complexity) doesn't need it. :)
+
+The clearest way to introduce the new syntax is to simply use it in an example similar to how it will be used by Anycloud:
+
+```ln
+from @std/tcpserver import connection
+from @std/tcp import Connection, connect, addContext, chunk, Chunk, Context, read, write, connClose, close, connError, connTimeout
+
+on connection fn (conn: Connection) {
+  const tunnel = connect('localhost', 8088);
+  conn.addContext(tunnel);
+  tunnel.addContext(conn);
+}
+
+on chunk fn (ctx: Context<Connection>) {
+  const c: Chunk = ctx.conn.read();
+  ctx.context.write(c);
+}
+
+on connClose fn (ctx: Context<Connection>) {
+  ctx.context.close();
+}
+
+on connError fn (ctx: Context<Connection>) {
+  ctx.context.close();
+}
+
+on connTimeout fn (ctx: Context<Connection>) {
+  ctx.context.close();
+}
+```
+
+The TCP server is a separate module for the same reason as the http server. The presence of a TCP listener would trigger the AVM to automatically start listening to an appropriate port. (Whether or not that autoselection sticks around long-term is still up in the air.)
+
+The connection handler in this case triggers the construction of a TCP client connection to the specified host and port. Once constructed, the server's connection and the client connection are added to each other's contexts.
+
+And that's it. Once constructed, the AVM will keep it resident and emit events involving that connection. Those events get the `Context<T>` type, where the `T` is the type that was added to the context. One known vulnerability with this as-is is that the client and server context types need to match each other and the compiler will not be able to keep this completely safe. This could potentially be avoided by having completely separate events for client and server, but that also nearly doubles the size of the codebase so it is being avoided for now. This can also be worked around with an `Either<ServerType, ClientType>` that is checked by the handlers which is why this is not considered a huge problem.
+
+The `Context<T>` type has two properties, the `conn: Connection` that corresponds to the connection that the triggered the event, and the `context: T` property that is whatever value was attached to the connection, which in this case is another connection.
```

Comment:
This part is what has me thinking/hesitant the most. The context seems like completely generic metadata that can be attached to any struct and not necessarily just a `connection`? I'm not sure if we should take the `context` concept to one of the opposite ends: either allow some sort of or arbitrary metadata to be attached to all structs or make this a special property of connection wherein one can associate two connections with one another.

I also keep wondering if it's possible to achieve this in a simpler way, without contention, through some sort of struct composition but it is not very clear to me how it would work.

Reply:
So we can call it `TcpContext` if that makes you more comfortable? I didn't want to go all the way to a [Go Context type](https://golang.org/src/context/context.go?s=2460:6019#L52) (though I did get some inspiration there) because with that the values attached to the context are stored in a hashmap that requires constant unsafe casting from `interface {}` to work with, so instead I went with a generic type for a singular value, which you could make a `HashMap<string, whatever>` but tried to minimize the "blast radius" of casting here.

The only other way it could potentially work out is with `@std/datastore`, shoving Context-like values in there and then pulling them out where necessary, but since datastore is intended to be replicated across the cluster it would be a massive performance bottleneck for manipulating raw TCP streams to require a full TCP request/response under the hood to get the relevant metadata.

Some sort of "forced-local" datastore model could be a compromise there, but still seems like a lower performance than the connection itself holding onto the context that will be needed.

Reply:
> So we can call it `TcpContext` if that makes you more comfortable? I didn't want to go all the way to a [Go Context type](https://golang.org/src/context/context.go?s=2460:6019#L52) (though I did get some inspiration there) because with that the values attached to the context are stored in a hashmap that requires constant unsafe casting from `interface {}` to work with, so instead I went with a generic type for a singular value, which you could make a `HashMap<string, whatever>` but tried to minimize the "blast radius" of casting here.
> 
> The only other way it could potentially work out is with `@std/datastore`, shoving Context-like values in there and then pulling them out where necessary, but since datastore is intended to be replicated across the cluster it would be a massive performance bottleneck for manipulating raw TCP streams to require a full TCP request/response under the hood to get the relevant metadata.
> 
> Some sort of "forced-local" datastore model could be a compromise there, but still seems like a lower performance than the connection itself holding onto the context that will be needed.

ah that makes a lot of sense. particularly why we are not considering with a full on `Context type`. I agree using datastore or backing it up with a HashMap would be too much overhead.

I think calling it `TcpContext` might make it easier to follow. What about having a parent `ConnectionPair` that connects both. Or perhaps getting closer to the semantics of a proper channel where a `TcpChannel` has a `tx` and `rx`?

Reply:
Keeping the `Connection` an opaque type was a goal so we can also treat it as a pointer under the hood like we did for the HTTP server in the AVM. Splitting the object into `tx` and `rx` inner properties would be possible, but would be a fabrication here that would also introduce overheads we don't really need? `.read()`ing and `.write()`ing a `Connection` is still very clear for the end users.

Renaming `Connection` to `TcpChannel` would be alright by me, though, if it makes more sense to you that way? Then `TcpContext` would have `channel` and `context` properties?

Reply:
> Keeping the `Connection` an opaque type was a goal so we can also treat it as a pointer under the hood like we did for the HTTP server in the AVM. Splitting the object into `tx` and `rx` inner properties would be possible, but would be a fabrication here that would also introduce overheads we don't really need? `.read()`ing and `.write()`ing a `Connection` is still very clear for the end users.
> 

I see. We definitely don't want to introduce unnecessary overhead.

> Renaming `Connection` to `TcpChannel` would be alright by me, though, if it makes more sense to you that way? Then `TcpContext` would have `channel` and `context` properties?

I think that does make more sense in my head, but I am getting tired so let me sleep on it.

Reply:
I prefer the `TcpChannel` name - it feels odd having `Connection` as the name when a Connection in other languages/libraries has a lot more information/properties attached to it